File: src/dataset/restructure/code_sanitizer.py
import subprocess
import logging
import tree_sitter

from dataclasses import dataclass
from common.common_typedef import Captures, TSNode
from tree_sitter import Node, QueryError, Tree, Query, QueryCursor

from ...common.tree_sitter_parser import TreeSitterParser

logger = logging.getLogger(__name__)


@dataclass
class CodeSanitizer:
    """Encapsulates the end-to-end process of cleaning, repairing, and
    refactoring source code snippets from a JSON dataset."""

    def __post_init__(self) -> None:
        """Initializes variables and necessary folders."""

        self.processed_data: dict[int, dict] = {}
        self.last_used_language: str | None = None

    def remove_non_ascii(self, code: str) -> str:
        """Removes all non-ASCII characters from a string."""
        return code.encode("ascii", "ignore").decode("ascii")

    def remove_comments(self, code: str, tsp: TreeSitterParser) -> str:
        """Parses the code and removes all comment nodes.

        Params:
            code: str
                The soruce code comments needs to be removed from.
            lang_name: str
                Language of the code to be processed.
        Returns:
            str
                source code cleaned up from comments
        """

        tree: Tree = tsp.parse(code)

        comments: list[Node] = [
            node
            for node in tsp.traverse_tree(node=tree.root_node)
            if node.type == "comment"
        ]

        # Iterate backwards to avoid index shifting issues during replacement (in-place replacement)
        for comment_node in sorted(comments, key=lambda c: c.start_byte, reverse=True):
            code = code[: comment_node.start_byte] + code[comment_node.end_byte :]

        return code.lstrip()

    def validate_and_extract_body(self, code: str, tsp: TreeSitterParser) -> str:
        """
        Validates the snippet contains a desirable compound statement and
        extracts the code from that point onward, discarding leading garbage.

        Params
        ------
        code: str
            source code to be modified.
        parser: TreeSitter
            class implementing Tree-Sitter based utilities.

        Returns
        -------
        str:
            the cleaned code string if valid, otherwise None.
        """

        tree = tsp.parse(code)

        # This query specifically targets compound statements we want to keep.
        # - function_definition: Standard C/C++ functions.
        # - template_declaration: C++ templates wrapping functions or classes.
        # - namespace_definition: C++ namespaces.
        # - class_specifier: C++ classes (which contain methods).
        # - compound_statement: A standalone { ... } block. This is our key
        #   for catching old K&R C functions, where the type is implicit and
        #   Tree-sitter might see it as a declaration followed by a compound statement.
        if tsp.language_name == "cpp":
            query_str = """
            [
              (function_definition)
              (template_declaration)
              (namespace_definition)
              (compound_statement)
            ] @target
            """
        else:
            # C query is restricted to C constructs
            query_str = """[ 
                (function_definition)
                (compound_statement)
                (zend_vm_handler)
            ] @target"""

        captures: Captures = tsp.run_query_on_node(node=tree.root_node, query_str=query_str)
        target_nodes: list[Node] = captures.get("target", [])
        if not target_nodes: return ""

        # find the earliest valid construct
        first_node: Node = min(target_nodes, key=lambda n: n.start_byte)

        # <---- handling for K&R C functions ---->
        # If we found a standalone compound_statement, check if its parent is
        # the root or an ERROR at the root. This confirms it's not a nested block.
        if first_node.type == "compound_statement":
            parent: TSNode = first_node.parent
            if parent and parent.type not in ["translation_unit", "ERROR"]:
                return ""

        return code[first_node.start_byte:]

    def balance_directives(self, code: str, tsp: TreeSitterParser) -> str:
        """Balances preprocessor directives using a text-based classification
        of fundamental `preproc_directive` nodes.

        Params:
            code: str
                source code to be modified.
        Returns:
            str: the restructured code
        """

        all_directives: list[tuple[Node, str]] = []
        tree = tsp.parse(code)

        for node in tsp.traverse_tree(node=tree.root_node):
            # ignore zero-width "phantom" nodes
            if node.start_byte == node.end_byte:
                continue

            node_text: bytes | None = node.text
            if node.type in {"#if", "#ifdef", "#ifndef"}:
                all_directives.append((node, "opener"))
            elif node.type in {"#endif", "preproc_directive"} and node_text:
                if node_text.startswith(b"#endif"):
                    all_directives.append((node, "closer"))

        all_directives.sort(key=lambda d: d[0].start_byte)
        balance = sum(1 if name == "opener" else -1 for _, name in all_directives)

        return code + ("\n#endif" * balance) if balance > 0 else code

    def add_missing_braces(self, code: str, tsp: TreeSitterParser) -> str:
        """
        Appends missing closing braces to a code snippet using a syntax-aware
        parser (tree-sitter) to avoid incorrectly counting braces within
        strings or comments.

        Params:
            code: The code snippet to check and fix.
            language: The tree-sitter language grammar to use (e.g., C or C++).

        Returns:
            The code snippet with missing braces appended.
        """

        tree: Tree = tsp.parse(code=code)
        query_string: str = """
            "{" @open
            "}" @close
        """
        captures: Captures = tsp.run_query_on_tree(tree=tree, query_str=query_string)
        openers = captures.get("open", [])
        closers = captures.get("close", [])

        def _filer_ghost_nodes(array: list[Node]):
            return [
                n
                for n in array
                if n.start_byte != n.end_byte and not (n.is_missing or n.is_error)
            ]

        open_braces: int = len(_filer_ghost_nodes(openers))
        close_braces: int = len(_filer_ghost_nodes(closers))
        balance = open_braces - close_braces
        if balance > 0:
            return code.rstrip() + "\n" + "}" * balance

        return code

    def add_missing_return_types(self, code: str, tsp: TreeSitterParser) -> str:
        """Finds and fixes C functions that are missing an explicit
        return type by adding the default 'int'.
        """

        is_broken_pattern: bool = False

        tree: Tree = tsp.parse(code=code)
        root_node: Node = tree.root_node

        if (
            root_node.child_count > 0
            and root_node.children[0].type == "function_definition"
        ):
            return code

        if root_node.child_count >= 2:
            # parsed as two top-level
            first_child = root_node.children[0]
            second_child = root_node.children[1]

            # Check for the two known broken patterns:
            # 1. (ERROR) followed by (compound_statement)
            # 2. (expression_statement) followed by (compound_statement)
            is_broken_pattern = (
                first_child.type in ["ERROR", "expression_statement"]
                and second_child.type == "compound_statement"
            )
        # case in which no type specifier and the function is parsed as macro_type_specifier
        if root_node and (
            (
                root_node.children[0].type == "declaration"
                and root_node.children[0].named_children[0].type
                == "macro_type_specifier"
            )
            or (root_node.children[0].type == "macro_type_specifier")
        ):
            is_broken_pattern = True

        if is_broken_pattern:
            return "int " + code

        return code

    def _find_identifier_in_declarator(self, node: Node) -> Node | None:
        """Recursively finds the base identifier node within a declarator."""

        if node.type == "identifier":
            return node
        declarator_child = node.child_by_field_name("declarator")
        if declarator_child:
            return self._find_identifier_in_declarator(declarator_child)

        return None

    def is_php_zend(self, code: str, tsp: TreeSitterParser) -> bool:
        tree: Tree = tsp.parse(code=code)
        root_node: Node = tree.root_node

        query_php_zend = """
        (translation_unit
            (zend_vm_handler) @zend
        )
        """
        try:
            captures: Captures = tsp.run_query_on_node(node=root_node, query_str=query_php_zend)
            if captures: return True  # applying this fix to the PHP Zend virtual machine would break the code
        except tree_sitter.QueryError as e: raise QueryError(f" ❌ Grammar is broken!\n {e} ")
        except Exception as e: raise e

        return False

    def kr_style_to_ansi(self, code: str, tsp: TreeSitterParser) -> str:
        """Transforms a K&R C-style function into a modern ANSI C function using
        a robust, two-stage Tree-sitter approach.
        """

        tree: Tree = tsp.parse(code=code)
        root_node: Node = tree.root_node


        # --- Stage 1: Normalization Pass ---
        # check for the "broken" AST pattern of a K&R function
        broken_kr_query_str = """
            (translation_unit
              (expression_statement) @header
              (declaration)+ @param_decls .
              (compound_statement) @body
            )
        """
        if tsp.run_query_on_node(node=root_node, query_str=broken_kr_query_str):
            code = "int " + code.strip()

        # --- Stage 2: Conversion Pass ---
        # all K&R functions should have a valid `function_definition` node.
        # This query finds a function_definition with K&R-style parameter declarations.
        kr_func_query_str = """
            (function_definition
              (storage_class_specifier)* @specifiers
              type: (primitive_type) @ret_type
              declarator: (function_declarator
                parameters: (parameter_list) @param_names)
              (declaration)+ @param_decls
              body: (compound_statement) @body
            ) @kr_function
        """

        captures: Captures = tsp.run_query_on_node(node=root_node, query_str=kr_func_query_str)

        param_names_nodes: list[Node] | None = captures.get("param_names", [])
        param_decl_nodes: list[Node] | None = captures.get("param_decls", [])

        if not (param_names_nodes and param_decl_nodes): return code

        param_names_node: Node = param_names_nodes[0]

        # --- Reconstruct the Parameter List ---
        param_names: list[str] = [
            p.text.decode("utf-8")
            for p in param_names_node.children
            if p.type == "identifier" and p.text
        ]

        param_decl_map: dict[str, str] = {}
        for decl_node in param_decl_nodes:
            type_node = decl_node.child_by_field_name("type")
            if not type_node or not type_node.text: continue

            type_text = type_node.text.decode("utf-8")

            for declarator_node in decl_node.children_by_field_name("declarator"):
                if not declarator_node.text: continue

                full_declarator_text = declarator_node.text.decode("utf-8")
                identifier_node = self._find_identifier_in_declarator(declarator_node)

                if identifier_node and identifier_node.text:
                    var_name = identifier_node.text.decode("utf-8")
                    param_decl_map[var_name] = f"{type_text} {full_declarator_text}"

        new_params = [param_decl_map.get(name, f"int {name}") for name in param_names]

        # --- Plan and Apply Edits ---
        edits = []
        new_param_list_text = f"({', '.join(new_params)})".encode("utf-8")
        edits.append((param_names_node.start_byte, param_names_node.end_byte, new_param_list_text))

        start_byte_to_remove = param_names_node.end_byte
        end_byte_to_remove = param_decl_nodes[-1].end_byte
        edits.append((start_byte_to_remove, end_byte_to_remove, b""))

        code_bytes = bytearray(code, "utf-8")
        for start, end, text in sorted(edits, key=lambda x: x[0], reverse=True):
            code_bytes[start:end] = text

        return code_bytes.decode("utf-8")

    def preprocess_code_gcc_e(self, code: str, gcc_flags: list | None = None) -> str:
        """Preprocesses a C code string using the external GCC preprocessor.

        This method invokes the `gcc -E -P -` command to run the C
        preprocessor on the input code string. It reads from stdin and
        captures the preprocessed output from stdout.

        Parameters
        ----------
        code : str
            A string containing the C source code to be preprocessed.
        gcc_flags : list of str, optional
            A list of optional flags to pass to the GCC preprocessor,
            e.g., `['-DDEBUG=1']` to define a macro. If None, no extra
            flags are used.

        Returns
        -------
        str
            The preprocessed C code as a string.

        Raises
        ------
        FileNotFoundError
            If the 'gcc' command is not found. Ensure GCC is installed and
            accessible in the system's PATH.
        subprocess.SubprocessError
            If the preprocessing command fails (returns a non-zero exit code),
            which can occur due to errors in the C code.
        """

        if gcc_flags is None:
            gcc_flags = []

        # Command to execute: gcc -E -P -
        # -E: run the preprocessor only.
        # -P: inhibit generation of linemarkers
        # -:  read from standard input.
        command: list[str] = ["gcc", "-E", "-P", "-"] + gcc_flags

        try:
            process = subprocess.run(
                command, input=code, text=True, capture_output=True, check=True
            )
            return process.stdout
        except FileNotFoundError:
            raise FileNotFoundError(" ❌ 'gcc' not found. Ensure it is installed and in Path")
        except subprocess.CalledProcessError as e:
            raise subprocess.SubprocessError(f" ❌ Error during preprocessing:\n{e.stderr}")

File: rusty/src/processor_lib/code_sanitizer.rs
use anyhow::{Context, anyhow};
use rayon;
use std::borrow::Cow;
use std::io::{self, Write};
use std::process::{Command, Stdio};
use tree_sitter::{Node, QueryCursor};
use tree_sitter::{StreamingIterator, Tree};
use which::which;

use crate::processor_lib::lazy::{
    BRACE_QUERY, COMMENT_QUERY, DIRECTIVE_BALANCE_QUERY, IS_CPP_QUERY, VALIDATION_QUERY,
};
use crate::processor_lib::lazy::{CPP_PATTERNS, LEADING_GARBAGE_PATTERN};
use crate::processor_lib::tree_sitter_parser::TreeSitterParser;

#[derive(Debug)]
pub struct CodeSanitizer;

impl CodeSanitizer {
    /// Removes all non-ASCII characters from a string.
    pub fn remove_non_ascii(&self, code: &str) -> String {
        code.chars().filter(|c| c.is_ascii()).collect()
    }

    /// Helper function to remove frequent garbage patterns from the start of the code.
    fn cleanup_leading_garbage<'a>(&self, code: &'a str) -> Cow<'a, str> {
        LEADING_GARBAGE_PATTERN.replace(code, "")
    }

    pub fn remove_comments<'a>(&self, code: &'a str, ts_parser: &TreeSitterParser) -> Cow<'a, str> {
        let code_after_comments = ts_parser.with_tree(code, |_parser, tree| {
            let mut cursor = QueryCursor::new();
            let mut matches = cursor.matches(&COMMENT_QUERY, tree.root_node(), code.as_bytes());

            // Collect the byte ranges (start, end) of the comments, not the Node objects.
            let mut comment_ranges: Vec<(usize, usize)> = Vec::new();

            while let Some(a_match) = matches.next() {
                let node = a_match.captures[0].node;
                comment_ranges.push((node.start_byte(), node.end_byte()));
            }

            if comment_ranges.is_empty() {
                return Cow::Borrowed(code);
            }

            // The rest of your logic now uses the collected ranges.
            let mut new_code = String::with_capacity(code.len());
            let mut last_end = 0;
            for (start_byte, end_byte) in comment_ranges {
                new_code.push_str(&code[last_end..start_byte]);
                last_end = end_byte;
            }
            new_code.push_str(&code[last_end..]);
            Cow::Owned(new_code)
        });

        let intermediate_cow = code_after_comments.unwrap_or(Cow::Borrowed(code));

        let cleaned_cow = self.cleanup_leading_garbage(&intermediate_cow);
        let final_str = cleaned_cow.trim_start();
        if final_str != code {
            Cow::Owned(final_str.to_string())
        } else {
            Cow::Borrowed(code)
        }
    }

    /// Validates that a code snippet contains a valid function-like construct
    /// and extracts the code from that point onward, discarding leading garbage.
    ///
    /// Returns a slice of the original code string if a valid construct is found, otherwise `None`.
    pub fn validate_and_extract_body<'a>(
        &self,
        code: &'a str,
        tree: &Tree,
    ) -> Option<Cow<'a, str>> {
        let mut cursor = QueryCursor::new();
        let mut matches_iterator =
            cursor.matches(&VALIDATION_QUERY, tree.root_node(), code.as_bytes());

        let mut first_node: Option<Node> = None;
        while let Some(query_match) = matches_iterator.next() {
            for capture in query_match.captures {
                let current_node = capture.node;

                match first_node {
                    Some(ref mut existing_node) => {
                        if current_node.start_byte() < existing_node.start_byte() {
                            *existing_node = current_node;
                        }
                    }
                    None => {
                        first_node = Some(current_node);
                    }
                }
            }
        }

        // After checking all matches, get the earliest node found.
        let earliest_node = first_node?;

        // Special handling for nested compound statements.
        if earliest_node.kind() == "compound_statement"
            && let Some(parent) = earliest_node.parent()
            && !["translation_unit", "ERROR"].contains(&parent.kind())
        {
            return None;
        }

        let start_byte = earliest_node.start_byte();
        if start_byte == 0 {
            Some(Cow::Borrowed(code))
        } else {
            Some(Cow::Owned(code[start_byte..].to_string()))
        }
    }

    pub fn balance_directives<'a>(
        &self,
        code: &'a str,
        ts_parser: &TreeSitterParser,
    ) -> Cow<'a, str> {
        let maybe_modified_code = ts_parser.with_tree(code, |_parser, tree| {
            let mut cursor = QueryCursor::new();
            let matches =
                cursor.matches(&DIRECTIVE_BALANCE_QUERY, tree.root_node(), code.as_bytes());
            let capture_names = DIRECTIVE_BALANCE_QUERY.capture_names();

            let balance = matches.fold(0, |acc, m| {
                for c in m.captures {
                    let name = capture_names[c.index as usize];
                    if name.starts_with('_') {
                        continue;
                    }
                    return match name {
                        "opener" => acc + 1,
                        "closer" => acc - 1,
                        _ => acc,
                    };
                }
                acc
            });

            if balance > 0 {
                let mut new_code = String::with_capacity(code.len() + (balance as usize * 7)); // "\n#endif" = 7 bytes
                new_code.push_str(code);
                for _ in 0..balance {
                    new_code.push_str("\n#endif");
                }
                Cow::Owned(new_code)
            } else {
                Cow::Borrowed(code)
            }
        });

        // If parsing failed (maybe_modified_code is None), return the original code.
        maybe_modified_code.unwrap_or(Cow::Borrowed(code))
    }

    pub fn add_missing_braces<'a>(&self, code: &'a str, tree: &Tree) -> Cow<'a, str> {
        let mut open_braces: usize = 0;
        let mut close_braces: usize = 0;
        let open_idx = BRACE_QUERY.capture_index_for_name("open").unwrap();
        let close_idx = BRACE_QUERY.capture_index_for_name("close").unwrap();

        let mut cursor = QueryCursor::new();
        let mut matches = cursor.matches(&BRACE_QUERY, tree.root_node(), code.as_bytes());

        while let Some(m) = matches.next() {
            for cap in m.captures {
                let node = cap.node;
                // filter out phantom, empty, or error nodes.
                if node.is_missing() || node.is_error() || node.start_byte() == node.end_byte() {
                    continue;
                }

                if cap.index == open_idx {
                    open_braces += 1;
                } else if cap.index == close_idx {
                    close_braces += 1;
                }
            }
        }

        let balance = open_braces.saturating_sub(close_braces);

        if balance > 0 {
            let trimmed_code = code.trim_end();
            let mut new_code = String::with_capacity(trimmed_code.len() + (balance * 2));
            new_code.push_str(trimmed_code);
            for _ in 0..balance {
                new_code.push_str("\n}");
            }
            Cow::Owned(new_code)
        } else {
            Cow::Borrowed(code)
        }
    }

    /// Finds and fixes C functions missing an explicit return type by prepending `int`.
    ///
    /// This method inspects the top-level nodes of the syntax tree for common patterns
    /// that tree-sitter produces when a return type is absent (e.g., an ERROR node
    /// followed by a compound statement).
    pub fn add_missing_return_types<'a>(&self, code: &'a str, tree: &Tree) -> Option<Cow<'a, str>> {
        let root_node = tree.root_node();
        let mut is_broken_pattern = false;

        let first_child = root_node.child(0)?;
        if first_child.kind() == "function_definition" {
            // happy path
            return Some(Cow::Borrowed(code));
        }

        // pattern 1: (ERROR) or (expression_statement) followed by a function body.
        if root_node.child_count() >= 2
            && let Some(second_child) = root_node.child(1)
            && ["ERROR", "expression_statement"].contains(&first_child.kind())
            && second_child.kind() == "compound_statement"
        {
            is_broken_pattern = true;
        }

        // Check for Pattern 2: The type is parsed as a macro.
        if !is_broken_pattern {
            if first_child.kind() == "macro_type_specifier" {
                is_broken_pattern = true;
            } else if first_child.kind() == "declaration"
                && let Some(named_child) = first_child.named_child(0)
                && named_child.kind() == "macro_type_specifier"
            {
                is_broken_pattern = true;
            }
        }

        if is_broken_pattern {
            let mut new_code = String::with_capacity(4 + code.len()); // "int " is 4 bytes
            new_code.push_str("int ");
            new_code.push_str(code);
            Some(Cow::Owned(new_code))
            // Some(Cow::Owned(format!("int {}", code)))
        } else {
            Some(Cow::Borrowed(code))
        }
    }

    // pub fn is_php_zend(&self, code: &str, tree: &Tree) -> bool {
    //     let mut cursor = QueryCursor::new();
    //     let mut matches = cursor.matches(&ZEND_QUERY, tree.root_node(), code.as_bytes());
    //     matches.next().is_some()
    // }

    pub fn call_gcc_preprocessor(
        &self,
        code: &str,
        flags: Option<&[&str]>,
    ) -> Result<String, String> {
        let mut command = Command::new("gcc");
        command
            .arg("-E") // run preprocessor
            .arg("-P") // no linemarkers
            .arg("-x") // force C
            .arg("c")
            .arg("-") // read stdin
            .stdin(Stdio::piped()) // don't expect keyboard input
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());

        // Add any extra flags if they were provided.
        if let Some(flags) = flags {
            command.args(flags);
        }

        let mut child = match command.spawn() {
            Ok(child) => child,
            Err(e) if e.kind() == io::ErrorKind::NotFound => {
                return Err(
                    "❌ 'gcc' not found. Ensure it is installed and in the system's PATH."
                        .to_string(),
                );
            }
            Err(e) => {
                return Err(format!("❌ Failed to spawn gcc process: {}", e));
            }
        };

        // The `?` here converts a potential io::Error into our error String.
        child
            .stdin // 1. Access the stdin handle
            .take() // 2. Take ownership of the handle
            .unwrap() // 3. Unwrap the Option
            .write_all(code.as_bytes()) // 4. Write the data
            .map_err(|e| format!("❌ Failed to write to gcc stdin: {}", e))?;

        // Wait for the process to finish and capture its output.
        let output = child
            .wait_with_output()
            .map_err(|e| format!("❌ Failed to wait for gcc process: {}", e))?;

        if output.status.success() {
            // On success, convert stdout to a String.
            String::from_utf8(output.stdout)
                .map_err(|e| format!("❌ GCC output was not valid UTF-8: {}", e))
        } else {
            // On failure, convert stderr to a String to show the error.
            let stderr = String::from_utf8_lossy(&output.stderr);
            Err(format!("❌ Error during preprocessing:\n{}", stderr))
        }
    }

    pub fn format_code(&self, code: &str) -> Result<String, anyhow::Error> {
        let clang_format_exe =
            which("clang-format").context("`clang-format` executable not found in PATH.")?;

        let mut child = Command::new(&clang_format_exe)
            .arg("-assume-filename=input.c") // avoids defaulting to C++ and conflict with
            // .clang-format
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .context(format!(
                "Failed to spawn clang-format at '{}'",
                clang_format_exe.display()
            ))?;

        let mut stdin = child
            .stdin
            .take()
            .expect("Failed to open child process stdin");
        let code_to_write = code.to_string();

        // handle I/O concurrently within Rayon's thread pool.
        let (write_result, output_result) = rayon::join(
            // write to stdin.
            move || stdin.write_all(code_to_write.as_bytes()),
            // wait for the process to finish while consuming stdout and stderr
            || child.wait_with_output(),
        );

        write_result.context("Failed to write to clang-format stdin")?;
        let output = output_result.context("Failed to wait for clang-format to exit")?;

        if output.status.success() {
            Ok(String::from_utf8(output.stdout)?)
        } else {
            let stderr = String::from_utf8_lossy(&output.stderr);
            Err(anyhow!(
                "clang-format failed with exit code {}:\n{}",
                output.status,
                stderr
            ))
        }
    }

    pub fn is_cpp(&self, code: &str, cpp_parser: &TreeSitterParser) -> bool {
        if CPP_PATTERNS.is_match(code) {
            return true;
        }

        // Slower, more accurate tree-sitter check using the safe wrapper.
        let query_matches = cpp_parser.with_tree(code, |_parser, tree| {
            let mut cursor = QueryCursor::new();
            let mut matches = cursor.matches(&IS_CPP_QUERY, tree.root_node(), code.as_bytes());
            matches.next().is_some()
        });

        query_matches.unwrap_or(false)
    }
}
